odl/trafos/fourier.py

```diff
@@ -1645,6 +1647,211 @@ def inverse(self):
+            one of the locations is not normalized in [-0.5; 0.5[
+        """
+        if not isinstance(non_uniform_samples, Iterable):
+            raise TypeError('`non_uniform_samples` is not iterable.')
+
+        if len(non_uniform_samples) == 0:
+            raise ValueError('`non_uniform_samples` is empty')
+
+        n_dim = len(self.shape)
+        if not all(len(sample) == n_dim for sample in non_uniform_samples):
+            raise ValueError('One of the samples in `non_uniform_samples` does'
+              ' not have the right dimension')
+
+        if not all(
+                self._is_sample_valid(sample) for sample in non_uniform_samples
+            ):
+            raise ValueError('One of the samples in `non_uniform_samples` is'
+              ' not a float between -0.5 and 0.5')
+
+    def _is_sample_valid(self, sample):
+        """Check whether a single sample is correct.
+
+        Parameters
+        ----------
+        sample : iterable
+            The sample whose location you want to check
+
+        Returns
+        -------
+        bool : Whether the sample contains only floats in [-0.5; 0.5[
+        """
+        floats = all(isinstance(coord, float) for coord in sample)
+        if floats:
+            normalized = all(-0.5 <= coord < 0.5 for coord in sample)
+        else:
+            normalized = False
+        return floats & normalized
+
+    def _normalize(self, x):
+        """Normalize the result of the non uniform FFT.
+
+        Parameters
+        ----------
+        x : `numpy.ndarray`
+            The results of the non uniform FFT
+
+        Returns
+        -------
+        out : `numpy.ndarray`
+            Normalized results of the non uniform FFT
+        """
+        out = x / np.sqrt(self.nfft.M)
+        return out
+
+    @property
+    def adjoint(self):
+        """Adjoint of this instance."""
+        if self.adjoint_class:
+            return self.adjoint_class(
+                shape=self.shape,
+                non_uniform_samples=self.non_uniform_samples,
+            )
+        raise NotImplementedError(
+            "Adjoint not implemented for this non-uniform fourier operator",
+        )
+
+
+class NonUniformFourierTransform(NonUniformFourierTransformBase):
+    """Forward Non uniform Fast Fourier Transform.
+    """
+    def __init__(self, shape, non_uniform_samples):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        shape : tuple
+            The dimensions of the data whose non uniform FFT you want to
+            compute
+        non_uniform_samples : iterable
+            List of the fourier space positions where the coefficients are
+            computed. Each position must be in [-0.5; 0.5[
+        """
+        super(NonUniformFourierTransform, self).__init__(
+            shape=shape,
+            non_uniform_samples=non_uniform_samples,
+            domain=discr_sequence_space(shape, dtype=np.complex128),
```

Comment:
This is too restrictive. It should be possible to use this operator with any `DiscreteLp` type domain, perhaps even with `cn` type spaces. Also other `dtype`s should be supported, even if the underlying library only supports double precision. In particular real-to-complex transforms should work. They don't have to be optimized but at least run.

My suggestion is that as with other operators, `domain` (for the forward transform) or `range` (for the adjoint) are parameters to `__init__`, and the range is inferred from that and the number of samples.

Further, I think the range should simply be `cn(num_samples)`, not a `DiscreteLp` The former is more general and reflects more the idea that the samples are unstructured.

Reply:
Ok I will try to look a bit more into that.
Like I said I was not sure I really understood how to play with the spaces. If you know where to start that would be cool.
I basically understand that `cn` is the space of complex tensors. But aside from that it's a bit shdady (and I just tried to replicate what I saw in other operators simplified).

However, isn't it a burden that the user has to figure which domain to use? I see that it's what has been done so far so let's not change that now. I am asking more out of curiosity for why it has been done like this.

Reply:
> However, isn't it a burden that the user has to figure which domain to use? I see that it's what has been done so far so let's not change that now. I am asking more out of curiosity for why it has been done like this.

It's a bit of a burden, I agree. But what you get in return is, for instance, a handle on the discretization that tells you exactly how to normalize frequency samples. I'd argue that's a net win. With additional helper functions that tell people what frequency range corresponds to their discretized function space, this operator becomes easier to user IMO. Apart from that, the space is usually dictated by the application anyway, so there's often not an additional step to be taken.
If you look through the [`examples`](https://github.com/odlgroup/odl/tree/master/examples) you will see that basically all of them start by defining a space (e.g., where images live), followed by code that just makes use of that. It's a very common pattern.

> I basically understand that `cn` is the space of complex tensors. But aside from that it's a bit shdady (and I just tried to replicate what I saw in other operators simplified).

For our case, the simplest variant of it applies, namely the space of N-long complex vectors. That's because the result of NFFT is basically just a list of complex numbers, corresponding to the Fourier samples at the chosen locations.
`DiscreteLp`, on the other hand, implies spatial structure: some kind of domain with a (rectangular) grid on it, where functions are discretized. We don't have that kind of structure here.

> Like I said I was not sure I really understood how to play with the spaces. If you know where to start that would be cool.

No worries, that's usually the hardest part to wrap your head around :-)
A good starting point would be this:

- Add a `space` as an argument and ensure it's a `DiscreteLp`. Also make sure it's uniformly discretized (`space.is_uniform`).
- Remove `shape` and take it from `space` instead.
- Take the per-axis `cell_sides` to compute the max frequencies.
- As for the samples, convert them to `np.ndarray` of a reasonable `dtype`. Check that their shape makes sense.
- Add a function for normalization of the frequencies based on the above computed max frequencies. Ideally, that computation would not run during `__init__` but during the first call to the operator, caching the normalized frequencies.

That should already take you quite far.

Reply:
Ok I see exactly what you wanted me to do and I think I have implemented it.
However, I didn't take `cell_sides` but `min_pt` and `max_pt` which, I think, are way better suited in this case. I just don't know if they are deprecated.

The examples did help understand what was going on.